### PR TITLE
New modular metric interface

### LINKED ISSUE #3069: metrics.sklearns don't work with DDP: Tensors must be CUDA and dense
## 🐛 Bug

Using sklearn-like metrics (from metrics.sklearns) works fine in DP mode, but results in the following error in DDP mode: `Tensors must be CUDA and dense`

### To Reproduce

This colab notebook has a full example: https://colab.research.google.com/drive/1uxihs4pztF_qTN-gkO1QOaX0npzE_uSZ?usp=sharing

Mentioning @justusschock

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Fixes #3069
   
This is a proposal on how an extension for the modular interface for
metric packages could look like. What our interface is missing, is the option to do computations
after dpp sync. Consider the following example for rmse:
    
    20 samples, 2 devices, 10 for each devices
    With our current setup, each device would return a metric value which then
    get synced.
    
    First machine returns (assume sum of squared error is 200)
    sqrt(1/10 * 200) = 4.47
    Second machine return (assume sum of squared error is 100)
    sqrt(1/10 * 100) = 3.16
    After ddp sync the value that is returned would be
    (4.47 + 3.16) / 2 = 3.815
    But the correct value is:
    sqrt(1/20 * 300) = 3.872
    
    It is possible to arrive at the correct result if each  machine instead returns 
    a *metric state* i.e a collection of values that can be synced and the correct results can
    derived from (in above case that would be the number of samples and the
    sum of squared errors). Thus, we need to extend the modular interface, such
    that we (or the user) can do calculations after the ddp sync.
    
This PR therefore propose to go from our decorator orientated modular interface
to a hook-based interface. All hooks are optional, such that the user only needs
to implement `forward` if the inherent from either `TensorMetric` or `NumpyMetric`.

Hooks to add:
* input_convert -> used to convert the input to correct type
* output_convert -> used to convert the output of forward
* ddp_sync -> do ddp_sync
* compute -> this is the missing part, that enables us to do computations after ddp sync
 
Note: this PR just implement the hooks, but I still need to go over each metric,
fixing those where the `compute` hook is needed.
That will be fixed in follow up PR, since this is already extensive enough.

Tagging @justusschock and @Borda for opinion

# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/metrics/metric.py [hunk 1]
```diff
@@ -17,6 +19,16 @@ class Metric(DeviceDtypeModuleMixin, torch.nn.Module, ABC):
...
     2. Handle their own DDP sync
+
+    Metric hooks that can be implemented are:
+        input_convert: pre-forward hook that takes care of input conversion
+        output_convert: post-forward hook that takes care of output convertion
+        ddp_sync: implementation of ddp sync
+        compute: post-ddp sync for additional metric computations
+
+    Call order:
+        input_convert -> forward -> output_convert -> ddp_sync -> compute
```
Comment (justusschock) [inline pytorch_lightning/metrics/metric.py:30]: Regarding ddp_sync: Maybe this should just be called `aggregate` and can be used to aggregate different batches on the same node as well? The behaviour shouldn't be different from syncing between nodes/processes, right?

So maybe we can implement a proper aggregation method and don't reduce the outputs during ddp_sync, but just broadcast them and then call `aggregate` on this?

**Comment by SkafteNicki** [comment]
@justusschock thanks, I took me a while to figure out how to keep this as close to native pytorch as possible, but still expressive enough to support the features we need. 
I think the first step is to change the backend to accommodate the features that we need and then do the actual implementation of the features.

For aggregation over multiple batches, one way to achieve this to introduce a new `Accumulate` class that inserts a hook just before `ddp_sync`, that accumulate the states:
```
class Accumulate(nn.Module):
    def __init__(self, base_metric):
        super().__init__()
        self.base_metric = base_metric
        self.base_metric._forward_hooks.popitem(-1) # remove compute hook
        self.base_metric.register_forward_hook(self.save_state) # add save_state hook
        self.base_metric.register_forward_hook(self.base_metric.compute) # re-add compute hook
        self.state = None
    def forward(self, *args, **kwargs):
        return self.base_metric(*args, **kwargs)
    def save_state(self, module, input, output):
        if self.state is None:
            self.state = output
        else:
            state = [s+o for s,o in zip(self.state, output)]
            self.state = state
            output = state
        return output
    def reset(self):
        self.state = None
# Use case
metric = Accumulate(Metric(...))
for (data, target) in dataloader:
    pred = model(data)
    val = metric(pred, target) # val will be the accumulated value
```

**Comment by justusschock** [comment]
@SkafteNicki I see your point. But isn't it basically the same whether you accumulate across nodes or across different batches in the same node? This would probably avoid some code duplication. 

We should have a chat in slack considering the perfect integration, once we finished the API for metrics.

**Comment by SkafteNicki** [comment]
@justusschock it probably is the same syncing between nodes and across the same node, so I agree that it should somewhat be handled in the same way. Only difference is that accumulation on the same node, should probably be a feature that the user can enable/disable (i.e. could be a `accumulate` flag in the metric `__init__`). 

For now, I will rename the `ddp_sync` hook to `aggregate` as you proposed, and it will still just support sync between nodes. Then we can discuss on slack the remaining parts.

**Comment by justusschock** [comment]
@SkafteNicki How is it going with this PR?

**Comment by SkafteNicki** [comment]
@justusschock I have update the PR with a test, so this PR (also) seems to solve the pickle issues we currently have with the metrics (https://github.com/PyTorchLightning/pytorch-lightning/issues/2848). Should we try to get this merged within the next couple of days?

**Comment by justusschock** [comment]
@SkafteNicki definitely. I'll just give you a first review :)

File: pytorch_lightning/metrics/metric.py [hunk 2]
```diff
@@ -27,18 +39,90 @@ def __init__(self, name: str):
...
         Implements the actual metric computation.
 
         Returns:
-            metric value
+            metric value or metric state
 
         """
         raise NotImplementedError
 
+    def compute(self, module: nn.Module, data: Any, output: Any):
```
Comment (justusschock) [inline pytorch_lightning/metrics/metric.py:62]: why do you need the metric module? Shouldn't this be available through `self?`

File: pytorch_lightning/metrics/metric.py [hunk 3]
```diff
@@ -27,18 +39,90 @@ def __init__(self, name: str):
...
+        """
+        Implement additionally metric computations to be done after the ddp sync
+
+        Args:
+            module: current metric module
+
+            data: input to forward method
+
+            output: output from forward method
```
Comment (justusschock) [inline pytorch_lightning/metrics/metric.py:71]: where do you get the output from? Shouldn't this be just computed in here?

File: pytorch_lightning/metrics/metric.py [hunk 4]
```diff
@@ -27,18 +39,90 @@ def __init__(self, name: str):
...
+
+        Returns:
+            final metric value
+
+        """
+        return output
+
+    def ddp_sync(self, module: nn.Module, data: Any, output: Any):
```
Comment (justusschock) [inline pytorch_lightning/metrics/metric.py:79]: Same as above

File: pytorch_lightning/metrics/metric.py [hunk 5]
```diff
@@ -27,18 +39,90 @@ def __init__(self, name: str):
...
+
+            output: output from forward method
+
+        Returns:
+            synced output
+
+        """
+        return output
+
+    def input_convert(self, module: nn.Module, data: Any):
```
Comment (justusschock) [inline pytorch_lightning/metrics/metric.py:96]: same as above

File: pytorch_lightning/metrics/metric.py [hunk 6]
```diff
@@ -27,18 +39,90 @@ def __init__(self, name: str):
...
+            module: current metric module
+
+            data: input to forward method
+
+        Returns:
+            casted data
+        """
+        return data
+
+    def output_convert(self, module: nn.Module, data: Any, output: Any):
```
Comment (justusschock) [inline pytorch_lightning/metrics/metric.py:110]: also see above

File: pytorch_lightning/metrics/metric.py [hunk 7]
```diff
@@ -27,18 +39,90 @@ def __init__(self, name: str):
...
+
+            data: input to forward method
+
+            output: output from forward method
+
+        Returns:
+            casted outputs
+        """
+        return output
+
```
Comment (justusschock) [inline pytorch_lightning/metrics/metric.py:125]: can we also add an aggregate function, that also aggregates samples on the same device?

File: tests/metrics/test_metrics.py [hunk 8]
```diff
@@ -139,3 +142,27 @@ def change_and_check_device_dtype(device, dtype):
...
         assert metric.dtype == torch.float16
         assert metric(input1, input2).dtype == torch.float16
+
+
+@pytest.mark.skipif(torch.cuda.device_count() < 2, reason="test requires multi-GPU machine")
+@pytest.mark.parametrize("metric", [
+    DummyTensorMetric,
+    DummyNumpyMetric(),
+])
+def test_pickable(tmpdir, metric: Metric):
```
Comment (justusschock) [inline tests/metrics/test_metrics.py:152]: can we also add a test, that simply tries to store and load the metric with `torch.save` and `torch.load` as this is a more direct test for this?

File: pytorch_lightning/metrics/metric.py (hunk 2 above)
Comment (SkafteNicki) [reply pytorch_lightning/metrics/metric.py]: Yes, you are right. A forward hook always takes 3 arguments: the module, input to the forward method and output of the forward method (https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.register_forward_hook). So having both a self and module argument was redundant. I deleted one arg and made these functions as @staticmethods instead

File: pytorch_lightning/metrics/metric.py (hunk 3 above)
Comment (SkafteNicki) [reply pytorch_lightning/metrics/metric.py:88]: The output is whatever is returned by the forward method the user have implemented. Thus, if the user returns some inter median values (like number of samples and the squared error in my written example) these are reduced by the ddp_sync hook and when we reach this method we can to the final computations to get the correct value

File: tests/metrics/test_metrics.py (hunk 8 above)
Comment (SkafteNicki) [reply tests/metrics/test_metrics.py]: added :)

**Comment by justusschock** [comment]
@SkafteNicki could you quickly summarise, what's still missing here?

**Comment by SkafteNicki** [comment]
I am not sure if anything are really missing. Status since last is:
* I have added the `aggregate` hook that you asked for and a function for gathering results instead of reducing them. However, none of these are actually used right now, since I could not figure out how to use them in a proper way
* It is important to remember that this PR is just a refactoring underneath, that only (directly) solves the pickle error we are having.
* That said, when this get merged, I will go over each and every metric to make sure that they are they are calculating the right stuff even in ddp (I plan on doing if metric individually such that the PRs are small and easier to get merged)

File: pytorch_lightning/metrics/metric.py [hunk 9]
```diff
@@ -41,18 +58,108 @@ def __init__(self, name: str):
...
+    @staticmethod
+    def output_convert(self, data: Any, output: Any):
+        """
+        Implement how outputs from forward should be casted
+
+        Args:
+
+            data: input to forward method
+
+            output: output from forward method
```
Comment (rohitgr7) [inline pytorch_lightning/metrics/metric.py:105]: are these blank lines required between the args docs?

File: pytorch_lightning/metrics/converters.py [hunk 10]
```diff
@@ -86,28 +86,32 @@ def new_func(*args, **kwargs):
...
+def convert_to_tensor(data: Any, dtype=None, device=None) -> Any:
     """
     Maps all kind of collections and numbers to tensors.
 
     Args:
         data: the data to convert to tensor
 
+        dtype: data type to convert to
+
+        device: device to cast to
```
Comment (rohitgr7) [inline pytorch_lightning/metrics/converters.py:98]:
```suggestion
        data: the data to convert to tensor
        dtype: data type to convert to
        device: device to cast to
```

File: pytorch_lightning/metrics/metric.py [hunk 11]
```diff
@@ -41,18 +58,106 @@ def __init__(self, name: str):
...
+        self.register_forward_hook(self.compute)
+
+    @staticmethod
+    def input_convert(self, data: Any):
+        """
+        Implement how the inputs should be casted before calling forward
+
+        Args:
+
+            data: input to forward method
```
Comment (rohitgr7) [inline pytorch_lightning/metrics/metric.py:78]:
```suggestion
            data: input to forward method
```

File: pytorch_lightning/metrics/metric.py [hunk 12]
```diff
@@ -41,18 +58,106 @@ def __init__(self, name: str):
...
+    @staticmethod
+    def ddp_sync(self, data: Any, output: Any):
+        """
+        Implement how the outputs from forward should be synced
+
+        Args:
+
+            data: input to forward method
+
+            output: output from the `output_convert` hook
```
Comment (rohitgr7) [inline pytorch_lightning/metrics/metric.py:119]:
```suggestion
            data: input to forward method
            output: output from the `output_convert` hook
```

File: pytorch_lightning/metrics/metric.py [hunk 13]
```diff
@@ -41,18 +58,106 @@ def __init__(self, name: str):
...
+    @staticmethod
+    def aggregate(self, data: Any, output: Any):
+        """
+        Implement aggregation of values on the same device
+
+        Args:
+
+            data: input to forward method
+
+            output: output from the `ddp_sync` hook
```
Comment (rohitgr7) [inline pytorch_lightning/metrics/metric.py:136]:
```suggestion
            data: input to forward method
            output: output from the `ddp_sync` hook
```

File: pytorch_lightning/metrics/metric.py [hunk 14]
```diff
@@ -41,18 +58,106 @@ def __init__(self, name: str):
...
+    @staticmethod
+    def compute(self, data: Any, output: Any):
+        """
+        Implement additionally metric computations to be done after the ddp sync
+
+        Args:
+
+            data: input to forward method
+
+            output: output from the `aggregate` hook
```
Comment (rohitgr7) [inline pytorch_lightning/metrics/metric.py:153]:
```suggestion
            data: input to forward method
            output: output from the `aggregate` hook
```

File: tests/metrics/test_metrics.py [hunk 15]
```diff
@@ -139,3 +143,45 @@ def change_and_check_device_dtype(device, dtype):
...
+    x, y = torch.randn(10,), torch.randn(10,)
+    results_before_save = metric(x,y)
+
+    # save metric
+    save_path = os.path.join(tmpdir, 'save_test.ckpt')
+    torch.save(metric, save_path)
+
+    # load metric
+    new_metric = torch.load(save_path)
+    results_after_load = new_metric(x,y)
```
Comment (rohitgr7) [inline tests/metrics/test_metrics.py:184]:
```suggestion
    x, y = torch.randn(10,), torch.randn(10,)
    results_before_save = metric(x, y)

    # save metric
    save_path = os.path.join(tmpdir, 'save_test.ckpt')
    torch.save(metric, save_path)

    # load metric
    new_metric = torch.load(save_path)
    results_after_load = new_metric(x, y)
```

File: CHANGELOG.md [hunk 16]
```diff
@@ -55,6 +55,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Added warning when changing monitor and using results obj ([#3014](https://github.com/PyTorchLightning/pytorch-lightning/pull/3014))
 - Added a hook `transfer_batch_to_device` to the `LightningDataModule` ([#3038](https://github.com/PyTorchLightning/pytorch-lightning/pull/3038))
 
+- Added hooks to metric module interface ([#2528](https://github.com/PyTorchLightning/pytorch-lightning/pull/2528/))
+
```
Comment (rohitgr7) [inline CHANGELOG.md:59]: since 0.9 is already released. I guess this should be moved up.

File: pytorch_lightning/metrics/metric.py [hunk 17]
```diff
@@ -74,15 +172,25 @@ def __init__(self, name: str,
...
+                                   (torch.Tensor, np.ndarray, numbers.Number),
+                                   convert_to_tensor,
+                                   self.dtype, self.device)
 
-    def __call__(self, *args, **kwargs) -> torch.Tensor:
-        def _to_device_dtype(x: torch.Tensor) -> torch.Tensor:
-            return x.to(device=self.device, dtype=self.dtype, non_blocking=True)
+    @staticmethod
+    def output_convert(self, data: Any, output: Any):
+        return apply_to_collection(output, torch.Tensor, convert_to_tensor,
```
Comment (Borda) [inline pytorch_lightning/metrics/metric.py:187]:
```suggestion
        return apply_to_collection(output, (torch.Tensor, np.ndarray, numbers.Number), convert_to_tensor,
```
shall it be the same as `TensorCollectionMetric`?

File: tests/base/model_train_steps.py [hunk 18]
```diff
@@ -175,3 +175,20 @@ def eval_epoch_end_full_loop_result_obj_dp(self, result):
         setattr(result, f'{eval_name}_step_end_metric', reduced)
 
         return result
+
+    def training_step_using_metrics(self, batch, batch_idx, optimizer_idx=None):
```
Comment (Borda) [inline tests/base/model_train_steps.py:179]:
```suggestion
    def training_step___using_metrics(self, batch, batch_idx, optimizer_idx=None):
```

File: tests/metrics/test_metrics.py [hunk 19]
```diff
@@ -139,3 +143,45 @@ def change_and_check_device_dtype(device, dtype):
         metric.half()
         assert metric.dtype == torch.float16
         assert metric(input1, input2).dtype == torch.float16
+
+
+@pytest.mark.skipif(torch.cuda.device_count() < 2, reason="test requires multi-GPU machine")
+@pytest.mark.parametrize("metric", [
+    DummyTensorMetric,
+    DummyNumpyMetric(),
```
Comment (Borda) [inline tests/metrics/test_metrics.py:151]: class or object here?

File: pytorch_lightning/metrics/metric.py (hunk 17 above)
Comment (SkafteNicki) [reply pytorch_lightning/metrics/metric.py:187]: `TensorMetric` assumes that everything is tensors whereas the `TensorMetricCollection` users can also output numpy arrays ect. In the future we should probably collaps these into one class (`TensorMetric` is restricted version of `TensorMetricCollection`).

**Review by Borda — APPROVED** [review]
LGTM 🐰

**Comment by Borda** [comment]
@SkafteNicki some metric test seems to be hanging...

**Comment by SkafteNicki** [comment]
@Borda which test are hanging, I cannot figure it out from drone details?

**Comment by Borda** [comment]
> @Borda which test are hanging, I cannot figure it out from drone details?

see this build - http://35.192.60.23/PyTorchLightning/pytorch-lightning/8999
the las passing tests was `tests/metrics/test_metrics.py::test_metric[metric1] PASSED  [ 37%]` so the next one hanged...

**Comment by SkafteNicki** [comment]
@Borda fixed the bug, we can merge this now :]
